File: golang/46.go
package golang
func permute(nums []int) [][]int {
    result := &[][]int{}
    used := make([]bool, len(nums))
    dfs(nums, result, []int{}, used)
    return *result
}

func dfs(nums []int, result *[][]int, current []int, used []bool) {
    if len(current) == len(nums) {
        *result = append(*result, append([]int{}, current...))
        return
    }
    for i:=0; i<len(nums); i++ {
        if used[i] {
            continue
        }
        used[i] = true
        current = append(current, nums[i])
        dfs(nums, result, current, used)
        current = current[:len(current) - 1]
        used[i] = false
    }
}

File: python/46.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        used = [False] * len(nums)
        result = []
        self.dfs(nums, used, result, [])
        return result

    def dfs(self, nums: List[int], used: list[bool], result: List[List[int]], current: List[int]):
        if len(nums) == len(current):
            result.append(current.copy())
            return

        for i in range(len(nums)):
            if used[i]:
                continue

            used[i] = True
            current.append(nums[i])

            self.dfs(nums, used, result, current)

            current.pop()
            used[i] = False